haystack/components/websearch/serpex.py

```diff
@@ -0,0 +1,210 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any, Literal, Optional
+
+import httpx
+from tenacity import retry, retry_if_exception_type, stop_after_attempt, wait_exponential
+
+from haystack import component, default_from_dict, default_to_dict, logging
+from haystack.dataclasses import Document
+from haystack.utils import Secret, deserialize_secrets_inplace
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class SerpexWebSearch:
+    """
+    Fetches web search results from the SERPEX API.
+
+    SERPEX provides web search results from multiple search engines including Google, Bing, DuckDuckGo, and more.
+    Use it to retrieve organic search results, snippets, and metadata for search queries.
+
+    ### Usage example
+
+    ```python
+    from haystack.components.websearch import SerpexWebSearch
+    from haystack.utils import Secret
+
+    fetcher = SerpexWebSearch(api_key=Secret.from_token("your-serpex-api-key"))
+    results = fetcher.run(query="What is Haystack?")
+
+    documents = results["documents"]
+    for doc in documents:
+        print(f"Title: {doc.meta['title']}")
+        print(f"URL: {doc.meta['url']}")
+        print(f"Snippet: {doc.content}")
+    ```
+    """
+
+    def __init__(
+        self,
+        *,
+        api_key: Secret = Secret.from_env_var("SERPEX_API_KEY"),
+        engine: Literal["auto", "google", "bing", "duckduckgo", "brave", "yahoo", "yandex"] = "google",
+        num_results: int = 10,
+        timeout: float = 10.0,
+        retry_attempts: int = 2,
+    ) -> None:
+        """
+        Initializes the SerpexWebSearch component.
+
+        :param api_key: SERPEX API key for authentication. Get yours at https://serpex.dev
+        :param engine: Search engine to use. Options: "auto", "google", "bing", "duckduckgo",
+                      "brave", "yahoo", "yandex". Defaults to "google".
+        :param num_results: Number of search results to return. Defaults to 10.
+        :param timeout: Timeout in seconds for the API request. Defaults to 10.0.
+        :param retry_attempts: Number of retry attempts for failed requests. Defaults to 2.
+        """
+        self.api_key = api_key
+        self.engine = engine
+        self.num_results = num_results
+        self.timeout = timeout
+        self.retry_attempts = retry_attempts
+
+        # Ensure that the API key is resolved
+        _ = self.api_key.resolve_value()
```

Comment:
I realize this is a pattern we are using in our other components, but we'd actually prefer to only run this `resolve_value` when we actually need the API key which is only at run time. So let's go ahead and remove this.